pkg/sql/schemachanger/scbuild/internal/scbuildstmt/dependencies.go

```diff
@@ -329,3 +331,19 @@ type NameResolver interface {
 	// ResolveConstraint retrieves a constraint by name and returns its elements.
 	ResolveConstraint(relationID catid.DescID, constraintName tree.Name, p ResolveParams) ElementResultSet
 }
+
+type DescriptorIDGenerator interface {
+	GenerateUniqueDescID() catid.DescID
+}
+
+// ReferenceProvider provides all referenced objects with in current DDL
+// statement. For example, CREATE VIEW and CREATE FUNCTION both could reference
+// other objects, and cross-references need to probably tracked.
+type ReferenceProvider interface {
+	// ForEachTableReference iterate through all referenced tables and the
+	// reference details with the given function.
+	ForEachTableReference(func(tbl catalog.TableDescriptor, refs []descpb.TableDescriptor_Reference) error) error
+	// ForEachTypeReference iterate through all referenced type ids with the given
+	// function.
+	ForEachTypeReference(func(typeID descpb.ID) error) error
+}
```

Comment:
We never, ever want to reason in terms of descriptors in `scbuildstmt`. Only elements. Perhaps your interface should be in terms of those validate* functions that you defined that consume this ReferenceProvider?

Reply:
This is absolutely rings a bell! Thanks for pointing it out. These methods are also need for building cross references besides validation for defending cross-db references. I would just remove the reference to table, instead just count on `table id` and `table type (table/view/sequence)` since that's all we need at the moment.

Reply:
Done.